File: Coding Sample/MS.NET/Fundamentals/Language/OOP/gctest.cs
using System;

class SomeResource : IDisposable
{
	private string id;

	static SomeResource()
	{
		#if TESTING
		Console.WriteLine("SomeResource type initialized.");
		#endif
	}

	public SomeResource(string name)
	{
		id = name;

		#if TESTING
		Console.WriteLine("{0} resource acquired.", id);
		#endif
	}

	public string Id
	{
		get {return id;}
	}

	public void Consume()
	{
		if(id != null)
			Console.WriteLine("{0} resource consumed.", id);
	}

	public void Dispose()
	{

		#if TESTING
		Console.WriteLine("{0} resource disposed.", id);
		#endif

		id = null;		
		GC.SuppressFinalize(this);
	}

	~SomeResource()
	{
		#if TESTING
		Console.WriteLine("{0} resource finalized.", id);
		#endif

		id = null;
	}
}


static class Test
{
	private static void Run()
	{
		SomeResource b = new SomeResource("Second");
		b.Consume();
	}

	public static void Main()
	{
		SomeResource a = new SomeResource("First");
		a.Consume();

		Run();

		Console.WriteLine("Before garbage-collection {0} resource is in generation:{1}", a.Id, GC.GetGeneration(a));
		GC.Collect(); //forcing garbage-collection
		GC.WaitForPendingFinalizers();
		Console.WriteLine("After garbage-collection {0} resource is in generation:{1}", a.Id, GC.GetGeneration(a));

		SomeResource c = new SomeResource("Third");
		c.Consume();
		c.Dispose();

		using(SomeResource d = new SomeResource("Fourth"))
		{
			d.Consume();
		}
	}
}

File: Coding Sample/Java/Fundamentals/Language/OOP/GCTest.java
class SomeResource implements AutoCloseable{

	private String id = null;

	static{
		System.out.println("SomeResource class initialized.");
	}

	SomeResource(String name){
		id = name;
		System.out.printf("%s resource acquired%n", id);
	}

	public void consume(){
		if(id != null)
			System.out.printf("%s resource consumed%n", id);
	}

	public void close(){
		if(id != null){
			System.out.printf("%s resource released%n", id);
			id = null;
		}
	}

	public void finalize(){
		close();
	}
}


class GCTest{

	private static void run(){
		SomeResource b = new SomeResource("Second");
		b.consume();
	}

	public static void main(String[] args){
		SomeResource a = new SomeResource("First");
		run();
		a.consume();
		a.close();
		System.gc();
		try(SomeResource c = new SomeResource("Third")){
			c.consume();
		}
	}
}

